File: app/src/main/java/com/bubelov/coins/ui/adapter/PlacesSearchResultsAdapter.java
package com.bubelov.coins.ui.adapter;

import android.content.Context;
import android.location.Location;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bubelov.coins.R;
import com.bubelov.coins.model.Place;
import com.bubelov.coins.util.DistanceUnits;
import com.bubelov.coins.util.DistanceUtils;
import com.google.android.gms.maps.model.LatLng;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author Igor Bubelov
 */

public class PlacesSearchResultsAdapter extends RecyclerView.Adapter<PlacesSearchResultsAdapter.ResultViewHolder> {
    private List<Place> places = new ArrayList<>();

    private Callback listener;

    private Location userLocation;

    private DistanceUnits distanceUnits;

    public PlacesSearchResultsAdapter(Callback listener, Location userLocation, DistanceUnits distanceUnits) {
        this.listener = listener;
        this.userLocation = userLocation;
        this.distanceUnits = distanceUnits;
    }

    @Override
    public ResultViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.list_item_places_search_result, parent, false);
        return new ResultViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ResultViewHolder holder, int position) {
        final Place place = places.get(position);
        holder.name.setText(place.getName());
        holder.ripple.setOnClickListener(v -> listener.onPlaceClick(place));

        if (userLocation == null) {
            holder.distance.setVisibility(View.GONE);
        } else {
            holder.distance.setVisibility(View.VISIBLE);

            Context context = holder.name.getContext();
            float distanceInKilometers = DistanceUtils.getDistance(new LatLng(place.getLatitude(), place.getLongitude()), userLocation) / 1000.0f;

            if (distanceUnits.equals(DistanceUnits.KILOMETERS)) {
                holder.distance.setText(context.getString(R.string.n_kilometers, distanceInKilometers));
            } else {
                float distanceInMiles = DistanceUtils.toMiles(distanceInKilometers);
                holder.distance.setText(context.getString(R.string.n_miles, distanceInMiles));
            }
        }

        holder.icon.setImageResource(R.drawable.ic_place);
    }

    @Override
    public int getItemCount() {
        return places.size();
    }

    public List<Place> getPlaces() {
        return places;
    }

    public void setPlaces(List<Place> places) {
        this.places = places;
    }

    static class ResultViewHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.ripple)
        View ripple;

        @BindView(R.id.icon)
        ImageView icon;

        @BindView(R.id.name)
        TextView name;

        @BindView(R.id.distance)
        TextView distance;

        ResultViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }

    public interface Callback {
        void onPlaceClick(Place place);
    }
}

File: app/src/main/java/com/bubelov/coins/ui/adapter/PlacesSearchResultsAdapter.kt
/*
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

package com.bubelov.coins.ui.adapter

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.bubelov.coins.R
import com.bubelov.coins.ui.model.PlacesSearchRow

import kotlinx.android.synthetic.main.row_places_search_result.view.*

class PlacesSearchResultsAdapter(
    private val items: List<PlacesSearchRow>,
    private val itemClick: (PlacesSearchRow) -> Unit
) : RecyclerView.Adapter<PlacesSearchResultsAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            LayoutInflater.from(parent.context).inflate(
                R.layout.row_places_search_result,
                parent,
                false
            )
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(items[position], itemClick)
    }

    override fun getItemCount() = items.size

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        fun bind(item: PlacesSearchRow, itemClick: (PlacesSearchRow) -> Unit) {
            itemView.apply {
                icon.setImageBitmap(item.icon)
                name.text = item.name
                distance.visibility = if (item.distance.isNotEmpty()) View.VISIBLE else View.GONE
                distance.text = item.distance
                setOnClickListener { itemClick(item) }
            }
        }
    }
}